File: data-pipeline-libraries-java/data-pipeline-core-java/src/main/java/com/enrichmeai/culvert/dataquality/DataQualityTransform.java
package com.enrichmeai.culvert.dataquality;

import com.enrichmeai.culvert.contracts.GovernancePolicy;
import com.enrichmeai.culvert.contracts.RuntimeContext;
import com.enrichmeai.culvert.contracts.StageMetrics;
import com.enrichmeai.culvert.contracts.Transform;
import com.enrichmeai.culvert.governance.MaskingPolicy;
import com.enrichmeai.culvert.governance.Masker;
import com.enrichmeai.culvert.schema.EntitySchema;
import com.enrichmeai.culvert.schema.SchemaField;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * A cloud-neutral {@link Transform} that validates each row against an
 * {@link EntitySchema} and classifies it as either a {@link ValidationResult.ValidRow}
 * or an {@link ValidationResult.InvalidRow} (with a full list of
 * {@link FieldViolation} instances — one per failing field).
 *
 * <h2>Validation rules (applied in order per field)</h2>
 * <ol>
 *   <li><strong>MISSING_REQUIRED</strong> — a field whose {@code mode} is
 *       {@code "REQUIRED"} has a {@code null} or absent value in the row map.</li>
 *   <li><strong>TYPE_MISMATCH</strong> — the field value is non-null but its
 *       runtime class is not assignable from the expected Java type for the wire
 *       type declared in the schema:
 *       <pre>
 *         STRING  → String
 *         INT64   → Long (or Integer, widened)
 *         FLOAT64 → Double (or Float, widened)
 *         BOOL    → Boolean
 *       </pre>
 *       Other wire types pass the type check (no known mapping).</li>
 *   <li><strong>OUT_OF_RANGE</strong> — the {@link SchemaField} carries a
 *       {@link NumericRange} (via {@link SchemaField#range()}) and the value,
 *       after conversion to {@code double}, falls outside {@code [min, max]}.
 *       This check is skipped if the value is null or already flagged with a
 *       type mismatch. No separate side-map is required — bounds live in the
 *       schema definition.</li>
 * </ol>
 *
 * <h2>All violations are accumulated</h2>
 * <p>A row with N failing fields produces an {@link ValidationResult.InvalidRow}
 * with N {@link FieldViolation} entries. Validation does not short-circuit on the
 * first failure.
 *
 * <h2>Usage — direct row validation</h2>
 * <pre>{@code
 * EntitySchema schema = EntitySchema.of("order", List.of(
 *         SchemaField.required("id",    "STRING"),
 *         SchemaField.required("amount","FLOAT64").withRange(NumericRange.of(0.0, 1_000_000.0)),
 *         SchemaField.nullable("note",  "STRING")));
 *
 * DataQualityTransform<Map<String,Object>> dq =
 *         new DataQualityTransform<>(schema, Function.identity());
 *
 * ValidationResult<Map<String,Object>> result = dq.validate(row);
 * if (result instanceof ValidationResult.ValidRow<Map<String,Object>> v) {
 *     downstreamSink.write(v.row());
 * } else if (result instanceof ValidationResult.InvalidRow<Map<String,Object>> inv) {
 *     deadLetterSink.write(inv);  // T14.2
 * }
 * }</pre>
 *
 * <h2>Usage — as a {@link Transform} in a pipeline stage</h2>
 * <pre>{@code
 * Transform<Map<String,Object>, ValidationResult<Map<String,Object>>> transform = dq;
 * Iterator<ValidationResult<Map<String,Object>>> results =
 *         transform.apply(inputIterator, runtimeContext);
 * }</pre>
 *
 * <p>The {@link #apply(Iterator, RuntimeContext)} implementation is lazy — it
 * wraps the input iterator with a mapping iterator and does not materialise the
 * full input. After the iterator is exhausted, one {@link StageMetrics} snapshot
 * is emitted via {@code context.stageMetrics()} with the total rows processed
 * and the count of invalid rows as errors.
 *
 * <h2>Zero cloud/Beam dependencies</h2>
 * <p>No {@code org.apache.beam.*} or {@code com.enrichmeai.culvert.gcp.*} imports.
 *
 * <h2>PII masking (opt-in, T14.4)</h2>
 * <p>An optional {@link GovernancePolicy} may be supplied at construction time.
 * When present, after a row passes all validation checks the transform iterates
 * the row's field map and applies masking to any field for which
 * {@link GovernancePolicy#maskingFor(String, String)} returns a non-empty policy.
 * The table name passed to {@code maskingFor} is {@code schema.name()}.
 * <p>Masking is a <em>post-validation</em> step: it runs only on
 * {@link ValidationResult.ValidRow} results, after all violation checks are
 * complete. Invalid rows are never masked.
 * <p>Backward-compat: when no policy is supplied (the original no-policy constructor),
 * masking is fully skipped — all existing tests are unaffected.
 * <p><strong>Mutable-map requirement:</strong> masking mutates the map returned
 * by the row accessor in place. Accessor implementations that return immutable
 * maps (e.g. {@code Map.of(...)}) will throw on {@code put}. Use a mutable
 * {@code HashMap} when masking is enabled.
 *
 * @param <R> The row type. The row-to-field-map accessor is supplied at
 *            construction time via a {@code Function<R, Map<String,Object>>}.
 *
 * @since Sprint 14 / issue #73 (T14.1); masking added T14.4 / issue #76;
 *        schema-grounded range validation T14.7 / issue #100
 */
public final class DataQualityTransform<R>
        implements Transform<R, ValidationResult<R>> {

    /** Maps schema wire types to expected Java types. */
    private static final Map<String, Class<?>> WIRE_TYPE_MAP = Map.of(
            "STRING",  String.class,
            "INT64",   Number.class,   // Long is Number; accept Integer widened too
            "FLOAT64", Number.class,   // Double and Float are both Number
            "BOOL",    Boolean.class
    );

    private final EntitySchema schema;
    private final java.util.function.Function<R, Map<String, Object>> rowAccessor;

    /**
     * Optional governance policy for PII masking. Null means masking is disabled
     * (backward-compatible default — the no-policy constructor leaves this null).
     */
    private final GovernancePolicy governancePolicy;

    // ---------------------------------------------------------------
    // Constructors
    // ---------------------------------------------------------------

    /**
     * Schema-only convenience constructor (no masking policy).
     *
     * <p>Range validation is driven entirely by
     * {@link SchemaField#range()} — no side-map required.
     *
     * @param schema      The schema to validate each row against.
     * @param rowAccessor Extracts a {@code Map<fieldName, value>} from a row.
     */
    public DataQualityTransform(
            EntitySchema schema,
            java.util.function.Function<R, Map<String, Object>> rowAccessor) {
        this(schema, rowAccessor, null);
    }

    /**
     * Full constructor with an optional governance policy for PII masking.
     *
     * <p>Range validation is driven by {@link SchemaField#range()} — no side-map
     * is accepted or required.
     *
     * <p>When {@code policy} is non-null, matched fields in passing rows are
     * masked in place (see class javadoc for the mutable-map requirement).
     *
     * @param schema      The schema to validate each row against.
     * @param rowAccessor Extracts a {@code Map<fieldName, value>} from a row.
     * @param policy      Optional governance policy for PII masking.
     *                    Pass {@code null} to disable masking (default).
     */
    public DataQualityTransform(
            EntitySchema schema,
            java.util.function.Function<R, Map<String, Object>> rowAccessor,
            GovernancePolicy policy) {
        this.schema           = Objects.requireNonNull(schema,      "schema must not be null");
        this.rowAccessor      = Objects.requireNonNull(rowAccessor, "rowAccessor must not be null");
        this.governancePolicy = policy; // null ⇒ masking disabled
    }

    // ---------------------------------------------------------------
    // Public API: per-row validation
    // ---------------------------------------------------------------

    /**
     * Validates a single row against the {@link EntitySchema}.
     *
     * <p>All violations are accumulated before returning; validation does not
     * short-circuit on the first failure.
     *
     * @param row The row to validate. Must not be null.
     * @return {@link ValidationResult.ValidRow} if no violations were found;
     *         {@link ValidationResult.InvalidRow} with all violations otherwise.
     */
    public ValidationResult<R> validate(R row) {
        Objects.requireNonNull(row, "row must not be null");
        Map<String, Object> fields = rowAccessor.apply(row);

        List<FieldViolation> violations = new ArrayList<>();

        for (SchemaField field : schema.fields()) {
            Object value = (fields != null) ? fields.get(field.name()) : null;

            // 1 — MISSING_REQUIRED
            if ("REQUIRED".equals(field.mode()) && value == null) {
                violations.add(new FieldViolation(
                        field.name(),
                        ViolationKind.MISSING_REQUIRED,
                        "Field '" + field.name() + "' is REQUIRED but was null or absent"));
                continue; // no point type-checking a null
            }

            if (value == null) {
                // nullable/repeated field with null value — no further checks
                continue;
            }

            // 2 — TYPE_MISMATCH
            Class<?> expected = WIRE_TYPE_MAP.get(field.type());
            boolean typeMismatch = false;
            if (expected != null && !expected.isInstance(value)) {
                violations.add(new FieldViolation(
                        field.name(),
                        ViolationKind.TYPE_MISMATCH,
                        "Field '" + field.name() + "' expected type compatible with "
                                + field.type() + " (" + expected.getSimpleName()
                                + ") but got " + value.getClass().getSimpleName()
                                + " [value=" + value + "]"));
                typeMismatch = true;
            }

            // 3 — OUT_OF_RANGE (schema-grounded: bounds come from SchemaField.range())
            if (!typeMismatch) {
                Optional<NumericRange> rangeOpt = field.range();
                if (rangeOpt.isPresent() && value instanceof Number num) {
                    NumericRange range = rangeOpt.get();
                    double d = num.doubleValue();
                    if (!range.contains(d)) {
                        violations.add(new FieldViolation(
                                field.name(),
                                ViolationKind.OUT_OF_RANGE,
                                "Field '" + field.name() + "' value " + d
                                        + " is outside range [" + range.min()
                                        + ", " + range.max() + "]"));
                    }
                }
            }
        }

        if (!violations.isEmpty()) {
            return new ValidationResult.InvalidRow<>(row, violations);
        }

        // ---- Post-validation PII masking (opt-in, T14.4) -----
        // Apply masking only when a GovernancePolicy is configured.
        // Invalid rows are never masked — the dead-letter path receives the
        // original values so violations can be diagnosed.
        if (governancePolicy != null && fields != null) {
            String tableName = schema.name();
            for (Map.Entry<String, Object> entry : fields.entrySet()) {
                Optional<MaskingPolicy> mp =
                        governancePolicy.maskingFor(entry.getKey(), tableName);
                if (mp.isPresent()) {
                    entry.setValue(Masker.mask(entry.getValue(), mp.get()));
                }
            }
        }

        return new ValidationResult.ValidRow<>(row);
    }

    // ---------------------------------------------------------------
    // Transform<R, ValidationResult<R>> implementation
    // ---------------------------------------------------------------

    /**
     * Lazily maps {@link #validate(Object)} over the input iterator.
     *
     * <p>The returned iterator is a thin wrapper; rows are validated on demand
     * as the caller calls {@code next()}. The input is not materialised.
     *
     * <p>After the iterator is exhausted, one {@link StageMetrics} snapshot is
     * emitted via {@code context.stageMetrics()}: {@code rowsProcessed} = total
     * rows seen; {@code errorCount} = count of invalid rows;
     * {@code stageLatencyMs} = elapsed wall-clock time from the first call to
     * {@code next()} to iterator exhaustion.
     *
     * @param records The input rows.
     * @param context The runtime context (used for {@code stageMetrics()} emission).
     * @return A lazy iterator of {@link ValidationResult} instances.
     */
    @Override
    public Iterator<ValidationResult<R>> apply(
            Iterator<R> records, RuntimeContext context) {
        Objects.requireNonNull(records, "records must not be null");
        Objects.requireNonNull(context, "context must not be null");
        return new ValidatingIterator(records, context);
    }

    // ---------------------------------------------------------------
    // Inner iterator
    // ---------------------------------------------------------------

    private final class ValidatingIterator implements Iterator<ValidationResult<R>> {

        private final Iterator<R>    source;
        private final RuntimeContext context;

        private long rowsProcessed = 0L;
        private long errorCount    = 0L;
        private long startNanos    = -1L;

        ValidatingIterator(Iterator<R> source, RuntimeContext context) {
            this.source  = source;
            this.context = context;
        }

        @Override
        public boolean hasNext() {
            boolean more = source.hasNext();
            if (!more) {
                emitMetrics();
            }
            return more;
        }

        @Override
        public ValidationResult<R> next() {
            if (startNanos < 0) {
                startNanos = System.nanoTime();
            }
            ValidationResult<R> result = validate(source.next());
            rowsProcessed++;
            if (!result.isValid()) {
                errorCount++;
            }
            // Emit after last row (when iterator now empty)
            if (!source.hasNext()) {
                emitMetrics();
            }
            return result;
        }

        /**
         * Emits a single {@link StageMetrics} snapshot.
         * Guard ensures we emit exactly once even if hasNext() and next() both
         * detect exhaustion.
         */
        private boolean emitted = false;

        private void emitMetrics() {
            if (emitted) return;
            emitted = true;
            long elapsedNanos = (startNanos < 0) ? 0L : System.nanoTime() - startNanos;
            double latencyMs  = elapsedNanos / 1_000_000.0;
            try {
                context.stageMetrics().recordStageMetrics(new StageMetrics(
                        context.pipelineId(),
                        context.runId(),
                        "DataQualityTransform",
                        rowsProcessed,
                        latencyMs,
                        errorCount));
            } catch (Exception ex) {
                // Advisory — never let metrics emission break the pipeline.
                // (Same swallow-on-error contract as StageMetricsHook Javadoc)
            }
        }
    }
}

File: data-pipeline-libraries/data-pipeline-core/src/data_pipeline_core/dataquality/data_quality_transform.py
"""DataQualityTransform — cloud-neutral Transform that validates rows.

Mirrors ``com.enrichmeai.culvert.dataquality.DataQualityTransform``
(Java Sprint 14 / issue #73; masking T14.4 / issue #76;
schema-grounded range validation T14.7 / issue #100).

Port: T18.2 / issue #118.

Design notes vs Java
--------------------
* Java uses a lazy inner-class iterator (``ValidatingIterator``); Python
  uses a generator function which is idiomatically equivalent.
* Python ``bool`` is a subclass of ``int`` which is a subclass of
  ``numbers.Number``.  The Java ``WIRE_TYPE_MAP`` keeps ``Boolean`` and
  ``Number`` disjoint (Java lines 112-117), so INT64/FLOAT64 checks
  exclude ``bool`` explicitly here.
* No Python ``Masker`` exists in this package yet (Java counterpart is
  ``com.enrichmeai.culvert.governance.Masker``; Java lines 247-256).
  The ``governance_policy`` parameter is accepted for API parity but
  masking *application* is a FLAG — see class docstring.
* ``SchemaField.range`` is an additive field added in T18.2 to
  ``data_pipeline_core.schema.entity.SchemaField`` (mirrors Java T14.7).
"""

from __future__ import annotations

import numbers
import time
from typing import (
    Any,
    Callable,
    Dict,
    Generic,
    Iterator,
    List,
    Mapping,
    Optional,
    TypeVar,
)

from data_pipeline_core.contracts.stage_metrics import StageMetrics
from data_pipeline_core.dataquality.field_violation import FieldViolation
from data_pipeline_core.dataquality.numeric_range import NumericRange
from data_pipeline_core.dataquality.validation_result import (
    InvalidRow,
    ValidRow,
    ValidationResult,
)
from data_pipeline_core.dataquality.violation_kind import ViolationKind
from data_pipeline_core.schema.entity import EntitySchema, SchemaField

# TYPE_CHECKING import to avoid runtime circular dependency on GovernancePolicy.
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from data_pipeline_core.contracts.runtime import RuntimeContext

R = TypeVar("R")

# ---------------------------------------------------------------------------
# Wire-type → expected Python type map.
# Mirrors WIRE_TYPE_MAP (Java lines 112-117):
#   STRING  → str
#   INT64   → numbers.Number  (but NOT bool — see module docstring)
#   FLOAT64 → numbers.Number  (but NOT bool)
#   BOOL    → bool
# ---------------------------------------------------------------------------
_WIRE_TYPE_MAP: Dict[str, type] = {
    "STRING":  str,
    "INT64":   numbers.Number,
    "FLOAT64": numbers.Number,
    "BOOL":    bool,
}


def _is_type_match(wire_type: str, value: Any) -> bool:
    """Return True iff ``value`` is compatible with ``wire_type``.

    Mirrors Java ``expected.isInstance(value)`` (Java line 210) with the
    additional guard that ``bool`` is excluded from INT64/FLOAT64 checks
    (Python-specific: ``bool`` is a subclass of ``int`` and therefore of
    ``numbers.Number``, but a boolean is not a valid numeric field value
    in the Java model).
    """
    expected = _WIRE_TYPE_MAP.get(wire_type)
    if expected is None:
        # Unknown wire type — no type check (mirrors Java's pass-through).
        return True
    if expected is bool:
        return isinstance(value, bool)
    # For INT64/FLOAT64: must be a Number but must NOT be bool.
    return isinstance(value, expected) and not isinstance(value, bool)


class DataQualityTransform(Generic[R]):
    """A cloud-neutral Transform that validates rows against an EntitySchema.

    Mirrors ``DataQualityTransform<R>`` (Java lines 108-359).

    Validation rules (applied per field, in order — mirrors Java
    lines 192-237):

    1. **MISSING_REQUIRED** — a ``REQUIRED`` field with a ``None`` /
       absent value (Java line 194).
    2. **TYPE_MISMATCH** — the runtime type is incompatible with the
       schema wire type (Java lines 208-219).
    3. **OUT_OF_RANGE** — the field's numeric value falls outside the
       :class:`~data_pipeline_core.dataquality.NumericRange` attached to
       the :class:`~data_pipeline_core.schema.entity.SchemaField`
       (schema-grounded, Java lines 222-236).

    All violations are accumulated; validation does not short-circuit on
    the first failure (Java lines 48-50).

    PII masking (FLAG):
        The ``governance_policy`` constructor parameter is accepted to
        mirror the Java API (``DataQualityTransform.java:161-168``).
        However, no Python ``Masker`` exists in this package.  When a
        policy is supplied the transform raises ``NotImplementedError``
        on the first valid row that has a matching masking policy — do
        not use in production until a ``Masker`` is ported.
        Java masking path: lines 247-256.

    Usage — direct row validation::

        schema = EntitySchema(
            name="order",
            fields=[
                SchemaField("id",     "STRING",  mode="REQUIRED"),
                SchemaField("amount", "FLOAT64", mode="REQUIRED",
                            range=NumericRange.of(0.0, 1_000_000.0)),
                SchemaField("note",   "STRING"),
            ],
        )
        dq: DataQualityTransform[Dict[str, Any]] = DataQualityTransform(
            schema=schema, row_accessor=lambda row: row
        )
        result = dq.validate(row)
        if result.is_valid():
            ...
        else:
            assert isinstance(result, InvalidRow)
            for v in result.violations:
                print(v)

    Usage — as a Transform in a pipeline stage::

        results = dq.apply(iter(rows), runtime_context)
    """

    def __init__(
        self,
        schema: EntitySchema,
        row_accessor: Callable[[R], Optional[Mapping[str, Any]]],
        governance_policy: Optional[Any] = None,
    ) -> None:
        """Create a DataQualityTransform.

        Mirrors both Java constructors (Java lines 141-168).

        Args:
            schema:            The EntitySchema to validate each row against.
                               Mirrors ``schema`` (Java line 119).
            row_accessor:      Callable that extracts a field-name → value
                               mapping from a row.  Mirrors ``rowAccessor``
                               (Java line 120).
            governance_policy: Optional GovernancePolicy for PII masking.
                               Pass ``None`` to disable masking (default).
                               Mirrors ``governancePolicy`` (Java line 125-127).
                               FLAG: masking application is not yet implemented
                               in Python (no Masker).
        """
        if schema is None:
            raise ValueError("schema must not be None")
        if row_accessor is None:
            raise ValueError("row_accessor must not be None")
        self._schema = schema
        self._row_accessor = row_accessor
        self._governance_policy = governance_policy  # None → masking disabled

    # -----------------------------------------------------------------------
    # Public API: per-row validation
    # -----------------------------------------------------------------------

    def validate(self, row: R) -> ValidationResult[R]:
        """Validate a single row against the EntitySchema.

        Mirrors ``DataQualityTransform.validate(R)`` (Java lines 184-259).

        All violations are accumulated; validation does not short-circuit.

        Args:
            row: The row to validate.  Must not be ``None``.

        Returns:
            :class:`ValidRow` if no violations were found;
            :class:`InvalidRow` with all violations otherwise.
        """
        if row is None:
            raise ValueError("row must not be None")

        fields: Optional[Mapping[str, Any]] = self._row_accessor(row)
        violations: List[FieldViolation] = []

        for sf in self._schema.fields:
            value = (fields.get(sf.name) if fields is not None else None)

            # 1 — MISSING_REQUIRED (Java lines 193-199)
            if sf.mode == "REQUIRED" and value is None:
                violations.append(FieldViolation(
                    field_name=sf.name,
                    violation_kind=ViolationKind.MISSING_REQUIRED,
                    detail=(
                        f"Field '{sf.name}' is REQUIRED but was None or absent"
                    ),
                ))
                continue  # no point type-checking a None (Java line 199)

            if value is None:
                # nullable/repeated field with None value — no further checks
                # (Java lines 202-205)
                continue

            # 2 — TYPE_MISMATCH (Java lines 208-219)
            type_mismatch = False
            if not _is_type_match(sf.type, value):
                expected = _WIRE_TYPE_MAP.get(sf.type)
                violations.append(FieldViolation(
                    field_name=sf.name,
                    violation_kind=ViolationKind.TYPE_MISMATCH,
                    detail=(
                        f"Field '{sf.name}' expected type compatible with "
                        f"{sf.type} ({expected.__name__ if expected else 'unknown'}) "
                        f"but got {type(value).__name__} [value={value!r}]"
                    ),
                ))
                type_mismatch = True

            # 3 — OUT_OF_RANGE (Java lines 222-236)
            # Schema-grounded: bounds come from SchemaField.range (T14.7 / T18.2).
            # Skipped when value is None or already flagged with TYPE_MISMATCH.
            if not type_mismatch:
                numeric_range: Optional[NumericRange] = sf.range  # type: ignore[attr-defined]
                if (
                    numeric_range is not None
                    and isinstance(value, numbers.Number)
                    and not isinstance(value, bool)
                ):
                    d = float(value)
                    if not numeric_range.contains(d):
                        violations.append(FieldViolation(
                            field_name=sf.name,
                            violation_kind=ViolationKind.OUT_OF_RANGE,
                            detail=(
                                f"Field '{sf.name}' value {d} is outside range "
                                f"[{numeric_range.min}, {numeric_range.max}]"
                            ),
                        ))

        if violations:
            return InvalidRow.of(row=row, violations=violations)

        # ---- Post-validation PII masking (opt-in) --------------------------
        # FLAG: No Python Masker exists yet (Java: Masker.mask(), lines 247-256).
        # The governance_policy param is stored for API parity.  If a policy
        # is supplied we raise NotImplementedError to prevent silent no-ops.
        # When a Masker is ported, replace this block with the real call.
        if self._governance_policy is not None and fields is not None:
            for field_name, val in fields.items():
                mp = self._governance_policy.masking_for(field_name, self._schema.name)
                if mp is not None:
                    raise NotImplementedError(
                        "PII masking is accepted by the DataQualityTransform API "
                        "(mirrors Java T14.4 / issue #76) but the Python Masker "
                        "has not been ported yet (T18.2 flag).  Port "
                        "data_pipeline_core.dataquality.masker.Masker first."
                    )

        return ValidRow(row=row)

    # -----------------------------------------------------------------------
    # Transform[R, ValidationResult[R]] implementation
    # -----------------------------------------------------------------------

    def apply(
        self,
        records: Iterator[R],
        context: "RuntimeContext",
    ) -> Iterator[ValidationResult[R]]:
        """Lazily map :meth:`validate` over the input iterator.

        Mirrors ``DataQualityTransform.apply(Iterator<R>, RuntimeContext)``
        (Java lines 281-287) and the inner ``ValidatingIterator``
        (Java lines 293-358).

        The returned generator is lazy — rows are validated on demand.
        After the iterator is exhausted, one :class:`StageMetrics`
        snapshot is emitted via ``context.stage_metrics.record_stage_metrics``.
        Metrics-emission errors are swallowed (mirrors Java lines 352-356).

        Metrics:
            ``rows_processed`` — total rows validated.
            ``error_count``    — count of :class:`InvalidRow` results.
            ``stage_latency_ms`` — wall-clock from first ``next()`` call
                                   to exhaustion (0 on empty input).

        Args:
            records: The input rows.  Must not be ``None``.
            context: Runtime context (for metrics emission).  Must not be ``None``.

        Returns:
            A lazy iterator of :class:`ValidationResult` instances.
        """
        if records is None:
            raise ValueError("records must not be None")
        if context is None:
            raise ValueError("context must not be None")
        return self._validating_generator(records, context)

    def _validating_generator(
        self,
        records: Iterator[R],
        context: "RuntimeContext",
    ) -> Iterator[ValidationResult[R]]:
        """Generator implementation of the validating iterator.

        Mirrors ``ValidatingIterator`` (Java lines 293-358).
        """
        rows_processed = 0
        error_count = 0
        start_ns: Optional[int] = None

        try:
            for record in records:
                if start_ns is None:
                    start_ns = time.perf_counter_ns()  # first row (Java line 319)
                result = self.validate(record)
                rows_processed += 1
                if not result.is_valid():
                    error_count += 1
                yield result
        finally:
            # Emit metrics exactly once after the iterator is exhausted
            # (mirrors Java emitMetrics(), lines 340-357).
            elapsed_ns = (time.perf_counter_ns() - start_ns) if start_ns is not None else 0
            latency_ms = elapsed_ns / 1_000_000.0
            try:
                context.stage_metrics.record_stage_metrics(
                    StageMetrics(
                        pipeline_id=context.pipeline_id,
                        run_id=context.run_id,
                        stage_name="DataQualityTransform",  # Java line 349
                        rows_processed=rows_processed,
                        stage_latency_ms=latency_ms,
                        error_count=error_count,
                    )
                )
            except Exception:  # noqa: BLE001
                # Advisory — never let metrics emission break the pipeline.
                # Mirrors Java catch(Exception ex) swallow (Java lines 352-356).
                pass
